internal/datanode/compactor.go

```diff
@@ -0,0 +1,549 @@
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datanode
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/milvus-io/milvus/internal/log"
+	"github.com/milvus-io/milvus/internal/storage"
+	"github.com/milvus-io/milvus/internal/types"
+
+	"github.com/milvus-io/milvus/internal/proto/commonpb"
+	"github.com/milvus-io/milvus/internal/proto/datapb"
+	"github.com/milvus-io/milvus/internal/proto/etcdpb"
+	"github.com/milvus-io/milvus/internal/proto/schemapb"
+
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+var (
+	errCompactionTypeUndifined = errors.New("compaction type undefined")
+	errIllegalCompactionPlan   = errors.New("compaction plan illegal")
+	errTransferType            = errors.New("transfer intferface to type wrong")
+	errUnknownDataType         = errors.New("unknown shema DataType")
+)
+
+type iterator = storage.Iterator
+
+type compactor interface {
+	compact() error
+	getPlanID() UniqueID
+}
+
+type compactionTask struct {
+	downloader
+	uploader
+	compactor
+	Replica
+	flushManager
+	allocatorInterface
+
+	dc   types.DataCoord
+	plan *datapb.CompactionPlan
+}
+
+// check if compactionTask implements compactor
+var _ compactor = (*compactionTask)(nil)
+
+func newCompactionTask(
+	dl downloader,
+	ul uploader,
+	replica Replica,
+	fm flushManager,
+	alloc allocatorInterface,
+	dc types.DataCoord,
+	plan *datapb.CompactionPlan) *compactionTask {
+	return &compactionTask{
+		downloader:         dl,
+		uploader:           ul,
+		Replica:            replica,
+		flushManager:       fm,
+		allocatorInterface: alloc,
+		dc:                 dc,
+		plan:               plan,
+	}
+}
+
+func (t *compactionTask) getPlanID() UniqueID {
+	return t.plan.GetPlanID()
+}
+
+func (t *compactionTask) mergeDeltalogs(dBlobs map[UniqueID][]*Blob, timetravelTs Timestamp) (map[UniqueID]Timestamp, *DelDataBuf, error) {
+
+	dCodec := storage.NewDeleteCodec()
+
+	var (
+		pk2ts = make(map[UniqueID]Timestamp)
+		dbuff = &DelDataBuf{
+			delData: &DeleteData{Data: make(map[UniqueID]UniqueID)},
+			tsFrom:  math.MaxUint64,
+			tsTo:    0,
+		}
+	)
+
+	for _, blobs := range dBlobs {
+		_, _, dData, err := dCodec.Deserialize(blobs)
+		if err != nil {
+			log.Warn("merge deltalogs wrong", zap.Error(err))
+			return nil, nil, err
+		}
+
+		for pk, ts := range dData.Data {
+			if Timestamp(ts) >= timetravelTs {
```

Comment:
even if the delete is over timeTravelTs, should we still put it into deltaMap pk2ts map?